Formula/libvirt.rb

```diff
@@ -47,6 +48,25 @@ class Libvirt < Formula
     depends_on "libtirpc"
   end
 
+  # Don't generate accelerator command line on macOS
+  #
+  # This makes it once again possible to use the
+  #
+  #   <qemu:commandline>
+  #     <qemu:arg value='-machine'/>
+  #     <qemu:arg value='q35,accel=hvf'/>
+  #   </qemu:commandline>
+  #
+  # workaround to enable hardware acceleration.
+  #
+  # Drop once proper HVF support is added to libvirt.
+  #
+  # https://gitlab.com/libvirt/libvirt/-/issues/147
+  patch do
+    url "https://gitlab.com/abologna/libvirt/-/commit/da138afc3609a92d473fddcffa54b2020759f986.diff"
```

Comment:
Has this been submitted upstream?

Reply:
It hasn't, because it would never be accepted in its current form.

The proper way to address the underlying need is to add native HVF support to libvirt - this is what https://gitlab.com/libvirt/libvirt/-/issues/147 tracks. There are WIP patches dating back a few years, but unfortunately the author has been unable to find time to revive them yet. Hopefully we'll see progress in the area at some point in the not too distant future.

What this patch does is, it grossly chops away a bit of libvirt in order to make it possible to keep using the pre-7.10.0 workaround even post-7.10.0. It's not pretty, but with vanilla 7.10.0 there is simply no way to enable hardware acceleration for VMs and I think that even an ugly, unupstreamable workaround is preferable to that.

Reply:
We don't normally accept patches that won't be merged upstream. On the other hand, this is a patch that comes from an upstream developer, so most of our concerns about patches don't really apply here.

If this isn't acceptable to merge upstream (even if just temporarily), why is it suitable for us to carry here? Is there some reason why we should want an otherwise identically-configured Homebrew build to differ from one that a user builds from source for themselves?

Reply:
> We don't normally accept patches that won't be merged upstream. On the other hand, this is a patch that comes from an upstream developer, so most of our concerns about patches don't necessarily apply here.

I absolutely get it, and as an upstream maintainer I also don't want you to be carrying downstream patches either :)

> If this isn't acceptable to merge upstream (even if just temporarily), why is it suitable for us to carry here? Is there some reason why we should want an otherwise identically-configured Homebrew build to differ from one that a user builds from source for themselves?

Upstream has to care about targeting multiple platforms, but the narrower scope of Homebrew makes it possible to safely apply this kind of hack. This patch also completely breaks the test qemuxml2argv test, which would be a complete showstopper for upstream but is no big deal for Homebrew as you don't run the test suite as part of the build anyway.

Unfortunately I don't think it's possible to avoid Homebrew builds behaving differently from builds made from source: the current upstream status makes it impossible to use hardware acceleration, and that's not changing until proper HVF support is added to libvirt. I'm hoping that this will happen within a few months/releases, but there's no guarantee that it won't take much longer than that, and in the meantime both the few people building from source and the many, many people using Homebrew packages are going to be inconvenienced.

Ultimately, while I dislike downstream-only patches as much as the next person, I think they unfortunately represent the lesser evil in this specific situation :(

Reply:
> Upstream has to care about targeting multiple platforms, but the narrower scope of Homebrew makes it possible to safely apply this kind of hack.

IIUC, this patch skips the `qemuBuildAccelCommandLine` call when `__APPLE__` is defined (e.g. on macOS), and has no effect otherwise. Which platforms are adversely affected by this? Does this break things for iOS users?

> [Y]ou don't run the test suite as part of the build anyway.

This is true, but that doesn't quite mean we would knowingly merge something that breaks a test suite. Why is the test failure not a concern on macOS?

Reply:
> IIUC, this patch skips the `qemuBuildAccelCommandLine` call when `__APPLE__` is defined (e.g. on macOS), and has no effect otherwise. Which platforms are adversely affected by this? Does this break things for iOS users?

We don't support iOS as a target platform, so that's not a concern. But our CI pipeline includes a job that builds libvirt and runs its test suite on macOS (see [here](https://gitlab.com/libvirt/libvirt/-/jobs/1920748487) for the latest run) and that would break if we merged this patch.

> This is true, but that doesn't quite mean we would knowingly merge something that breaks a test suite. Why is the test failure not a concern on macOS?

The test suite had been broken on macOS for years before being fixed somewhat recently; throughout that time, the Homebrew package for libvirt has existed and has worked just fine, so it doesn't seem like this should be a dealbreaker if the ability to keep using hardware acceleration is the outcome of the compromise.

Plus, as mentioned above, upstream runs the test suite on macOS for every commit, so breakage is very likely to get caught way before it can make its way into a release and thus the Homebrew package.

Look, I'm really not happy with the situation either, but it is what it is. The lack of proper HVF support in libvirt has been a problem for a long time, but until 7.10.0 at least there was a reasonably straightforward workaround available to users. Now that has changed, and we can either leave hardware acceleration completely broken until HVF support is implemented or carry a downstream patch in Homebrew to restore the status quo. I would love a better solution, but I unfortunately can't think of any :(